posthog/hogql/database/schema/exchange_rate.py

```diff
@@ -19,3 +23,154 @@ def to_printed_clickhouse(self, context):
 
     def to_printed_hogql(self):
         return "exchange_rate"
+
+
+def convert_currency_call(
+    amount: ast.Expr, currency_from: ast.Expr, currency_to: ast.Expr, timestamp: ast.Expr | None = None
+) -> ast.Expr:
+    args = [currency_from, currency_to, amount]
+    if timestamp:
+        args.append(timestamp)
+
+    return ast.Call(name="convertCurrency", args=args)
+
+
+def revenue_currency_expression(config: RevenueTrackingConfig) -> ast.Expr:
+    exprs = []
+    for event in config.events:
+        exprs.extend(
+            [
+                ast.CompareOperation(
+                    left=ast.Field(chain=["event"]),
+                    op=ast.CompareOperationOp.Eq,
+                    right=ast.Constant(value=event.eventName),
+                ),
+                ast.Field(chain=["events", "properties", event.revenueCurrencyProperty])
+                if event.revenueCurrencyProperty
+                else ast.Constant(value=None),
+            ]
+        )
+
+    if len(exprs) == 0:
+        return ast.Constant(value=None)
+
+    # Else clause, make sure there's a None at the end
+    exprs.append(ast.Constant(value=None))
+
+    return ast.Call(name="multiIf", args=exprs)
+
+
+def revenue_comparison_and_value_exprs(
+    event: RevenueTrackingEventItem,
+    config: RevenueTrackingConfig,
+    do_currency_conversion: bool = False,
+) -> tuple[ast.Expr, ast.Expr]:
+    # Check whether the event is the one we're looking for
+    comparison_expr = ast.CompareOperation(
+        left=ast.Field(chain=["event"]),
+        op=ast.CompareOperationOp.Eq,
+        right=ast.Constant(value=event.eventName),
+    )
+
+    # If there's a revenueCurrencyProperty, convert the revenue to the base currency from that property
+    # Otherwise, assume we're already in the base currency
+    # Also, assume that `base_currency` is USD by default, it'll be empty for most customers
+    if event.revenueCurrencyProperty and do_currency_conversion:
+        value_expr = ast.Call(
+            name="if",
+            args=[
+                ast.Call(
+                    name="isNull", args=[ast.Field(chain=["events", "properties", event.revenueCurrencyProperty])]
+                ),
+                ast.Call(
+                    name="toDecimal",
+                    args=[
+                        ast.Field(chain=["events", "properties", event.revenueProperty]),
+                        ast.Constant(value=10),
+                    ],
+                ),
+                convert_currency_call(
+                    ast.Field(chain=["events", "properties", event.revenueProperty]),
+                    ast.Field(chain=["events", "properties", event.revenueCurrencyProperty]),
+                    ast.Constant(value=(config.baseCurrency or CurrencyCode.USD).value),
```

Comment:
Nit: We have this fallback in two places here, it is not much yet, but I wonder if it will happen more. What I would usually do is create a `@property` to centralize the fallback handler, I see there are places in the codebase where we do that but I don't see it being done on `posthog/models/schema.py` so maybe you'll know how to approach this if you think it's valid :)

Reply:
Awesome question. There's also a slightly better way where we make this non-optional, and add a `@default` to our pydantic models and it will handle that for us. I'm considering changing that, I'm checking with @robbie-c in a separate PR if that's the best way to do it actually - it'll also help with the new config I'm adding to `RevenueConfig`

Reply:
Yeah, it just doesn't exist in JS, you have to satisfy the types yourself. I think this is only working when sending data from the backend to the frontend, though, not the opposite, and it's not doing what I want. I'm testing this in a separate PR and will report back in our Slack channel with what I found out